### PR TITLE
Ra dspdc 501 core logic

### PR DESCRIPTION
**Pull request by raaidbroad**
Implements the core methods needed for the ingester MVP (barring the api status method). A few todos left here or there as possible things to improve upon later once the code is reviewed and more stable. Initial tests have also been added. The liquibase files for the db are present as well, though I'm not 100% certain I have everything needed for it (it lets the tests run!).

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -67,3 +75,41 @@ lazy val `jade-client` = project
...
+      "org.testcontainers" % "postgresql" % testcontainersVersion
+    ).map(_ % Test),
+    dependencyOverrides := Seq(
+      "org.postgresql" % "postgresql" % postgresqlDriverVersion
+    )
+  )
+
+lazy val `core-migrations` = project
+  .in(file("./core/db-migrations"))
+  .settings(commonSettings)
```
Comment (danxmoran) [inline build.sbt]: Can you add the newline? Same for other files missing one

File: build.sbt [hunk 2]
```diff
@@ -67,3 +75,41 @@ lazy val `jade-client` = project
...
+      "org.scalatest" %% "scalatest" % scalaTestVersion,
+      "org.testcontainers" % "postgresql" % testcontainersVersion
+    ).map(_ % Test),
+    dependencyOverrides := Seq(
+      "org.postgresql" % "postgresql" % postgresqlDriverVersion
+    )
+  )
+
+lazy val `core-migrations` = project
+  .in(file("./core/db-migrations"))
```
Comment (danxmoran) [inline build.sbt:114]: This is missing the settings needed to run the migrations locally / as a Docker container. I'm ok leaving it as-is for now as long as we create a ticket to track it.

File: core/db-migrations/src/main/resources/changesets/01-init-jobs-and-requests-tables.xml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            </column>
+            <column name="path" type="varchar">
+                <constraints nullable="false"/>
+            </column>
+            <column name="table_name" type="varchar">
+                <constraints nullable="false"/>
+            </column>
+            <column name="submitted" type="smalldatetime"/>
+            <column name="completed" type="smalldatetime"/>
+            <column name="updated" type="smalldatetime"/>
```
Comment (danxmoran) [inline core/db-migrations/src/main/resources/changesets/01-init-jobs-and-requests-tables.xml]: Why `smalldatetime` for these?

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/ApiStatus.scala [hunk 4]
```diff
@@ -0,0 +1,3 @@
+package org.broadinstitute.monster.ingester.core
+
+case class ApiStatus()
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/ApiStatus.scala]: ?

If this is a TODO to fill in, can you add a comment describing its intended purpose? And also create a ticket to track the implementation.

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/DoobieInstances.scala [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    * Typeclass which can read `OffsetDateTime`s from Postgres.
+    *
+    * First reads values as a SQL timestamp, then converts the underlying
+    * instant into a datetime in UTC.
+    */
+  implicit val odtGet: Get[OffsetDateTime] = Get[Timestamp].tmap { ts =>
+    OffsetDateTime.ofInstant(ts.toInstant, ZoneId.of("UTC"))
+  }
+
+  implicit val odtPut: Put[OffsetDateTime] = Put[Timestamp].contramap { ts =>
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/DoobieInstances.scala:35]: Can you add scaladoc?

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+) {
+
+  import DoobieInstances._
+  import Constants._
+
+  private val logger = Slf4jLogger.getLogger[IO]
+  private implicit val logHandler: LogHandler = DbLogHandler(logger)
+
+  // need a method to run all steps of an ingest; returns the request ID to check the status of?
+  def ingest(ingestRequest: IngestData, datasetId: UUID): IO[UUID] = ???
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala]: Can this be filled in as part of this PR, or are there still open questions?

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    *
+    * @param datasetId Id of the dataset in the Repo to ingest data into.
+    * @param now the current time for timestamps.
+    * @return the Id of the newly added request.
+    */
+  def createRequest(datasetId: UUID, now: Instant): IO[UUID] = {
+    // create request id
+    val requestId = UUID.randomUUID()
+
+    // TODO we might want to consider hitting Jade's API to validate the datasetId before
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala:78]: 👍 to this

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    // inserting it in the requests table
+    val updateSql = List(
+      Fragment.const(s"INSERT INTO $RequestsTable (id, submitted, dataset_id) VALUES"),
+      fr"($requestId," ++ Fragment.const(timestampSql(now)) ++ fr", $datasetId)"
+    ).combineAll
+
+    // return id and add to requestTable
+    val transaction = for {
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala]: `for`-comprehension is a bit heavyweight for this, simpler to:
```scala
updateSql.update.withUniqueGeneratedKeys[UUID]("id").transact(dbClient)
```

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+            prefix,
+            tableName // TODO maybe include updated timestamp in the initialization?
+          )
+      }
+      // update table
+      for {
+        jobsAdded <- Update[(UUID, JobStatus, String, String)](
+          s"INSERT INTO $JobsTable (request_id, status, path, table_name) VALUES (?, ?, ?, ?)"
+        ).updateMany(jobs)
+      } yield (jobsAdded)
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala]: Same as above, will be cleaner as not-a-`for`-comprehension

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+          .unique
+        counts <- List(
+          Fragment.const(s"SELECT COUNT(*), status FROM $JobsTable"),
+          fr"WHERE request_id = $rId",
+          fr"GROUP BY status"
+        ).combineAll
+          .query[(Long, JobStatus)]
+          .to[List]
+      } yield {
+        models.RequestSummary(submittedTime, counts)
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala]:
```suggestion
        RequestSummary(submittedTime, counts)
```

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+      for {
+        statuses <- List(
+          Fragment.const(
+            s"SELECT jade_id, status, path, table_name, submitted, completed FROM $JobsTable"
+          ),
+          fr"WHERE request_id = $rId"
+        ).combineAll
+          .query[JobSummary]
+          .to[List]
+      } yield (statuses)
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala]: Doesn't need to be a `for`

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+    }
+  }
+
+  // TODO need a method for our API's status, return a status type for the api (new type)
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala:231]: Can you make a ticket for this (and include filling in the ApiStatus model)?

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+        Fragment.const(s"SELECT jade_id FROM $JobsTable"),
+        fr"WHERE jade_id IS NOT NULL",
+        fr"AND status = ${JobStatus.Running: JobStatus}",
+        fr"ORDER BY updated ASC LIMIT $limit"
+      ).combineAll
+        .query[UUID]
+        .to[List]
+
+      // hit jade job result endpoint for each job
+      statuses <- Async[ConnectionIO].liftIO(ids.traverse { id =>
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala]: `ids.traverse(jadeClient.jobStatus)`

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/models/RequestSummary.scala [hunk 14]
```diff
@@ -0,0 +1,17 @@
+package org.broadinstitute.monster.ingester.core.models
+
+import java.time.OffsetDateTime
+
+import io.circe.{Decoder, Encoder}
+import io.circe.derivation.{deriveDecoder, deriveEncoder}
+import org.broadinstitute.monster.ingester.jade.models.JobStatus
+
+case class RequestSummary(
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/models/RequestSummary.scala:15]: Scaladoc?

**Review by danxmoran — COMMENTED** [review]
Only looked at the main code for now

File: core/db-migrations/src/main/resources/changesets/01-init-jobs-and-requests-tables.xml (hunk 3 above)
Comment (raaidbroad) [reply core/db-migrations/src/main/resources/changesets/01-init-jobs-and-requests-tables.xml]: Whoops probably something I tried when I was struggling with a timestamp doobie issue, I'll change the type to timestamp

File: build.sbt (hunk 2 above)
Comment (raaidbroad) [reply build.sbt:114]: Is this something I can add quickly by just messing with the build file or does it require other work?

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala (hunk 7 above)
Comment (raaidbroad) [reply core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala:78]: I can create a lower-priority ticket to add this later

File: build.sbt (hunk 2 above)
Comment (danxmoran) [reply build.sbt:114]: It'd touch a few files if we copy-pasted from Transporter. I want to add the configs to the sbt-plugins project, so ok to leave it out for now.

File: core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+    *
+    * @param ingestData the list of paths and table names to ingest from/to.
+    * @param datasetId the dataset ID to ingest into.
+    * @return the UUID of the request created.
+    */
+  def ingest(ingestData: IngestData, datasetId: UUID): IO[UUID] = {
+    for {
+      now <- getNow
+      rId <- createRequest(datasetId, now)
+      _ <- initJobs(ingestData, rId)
```
Comment (danxmoran) [inline core/src/main/scala/org/broadinstitute/monster/ingester/core/IngestController.scala:57]: Ahhh, I see now. `createRequest` and `initJobs` should run within a single transaction for this to be safe. If you change both to return `ConnectionIO` then you can chain them here and call `transact` once.

File: core/src/test/scala/org/broadinstitute/monster/ingester/core/IngestControllerSpec.scala [hunk 16]
```diff
@@ -0,0 +1,368 @@
...
+  behavior of "IngestController"
+
+  // createRequest
+  it should "return a UUID for valid inputs" in withController(apiEmpty) {
+    (_, controller) =>
+      for {
+        now <- getNow
+        a <- controller.createRequest(dataset1Id, now)
+      } yield {
+        a
```
Comment (danxmoran) [inline core/src/test/scala/org/broadinstitute/monster/ingester/core/IngestControllerSpec.scala]: Should this be asserting something?

Comment (raaidbroad) [reply core/src/test/scala/org/broadinstitute/monster/ingester/core/IngestControllerSpec.scala]: Nope I'll remove it; that was my "is this thing even working" test which is captured in the test after it while asserting things

**Review by danxmoran — APPROVED** [review]

**Pull request merged**